### PR TITLE
Replace C++ UpdateTask with an empty dart version, loaded as a snapshot

### PR DESCRIPTION
**Pull request by mpcomplete**
This adds a new placeholder dart package `updater` in sky/packages. This is built into a snapshot and compiled into the Sky engine binary using the same mechanism as the Dart isolate.

I also added a SkyHeadless class, similar to SkyView, used for running Dart code without a view.

### CODE REVIEW

File: sky/engine/core/script/dart_controller.cc [hunk 1]
```diff
@@ -74,6 +74,7 @@ DartController::~DartController() {
 }
 
 void DartController::DidLoadMainLibrary(String name) {
+  LOG(INFO) << "DidLoadMain: " << name.toUTF8();
```
Comment (abarth) [inline sky/engine/core/script/dart_controller.cc]: We probably don't want to land this debugging log.

File: sky/engine/public/sky/sky_headless.cc [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+SkyHeadless::~SkyHeadless() {
+}
+
+void SkyHeadless::Init(const String& name) {
+  DCHECK(!dart_controller_);
+
+  dart_controller_ = adoptPtr(new DartController);
+  LOG(INFO) << "CreateHeadless: " << name.toUTF8();
```
Comment (abarth) [inline sky/engine/public/sky/sky_headless.cc]: We probably don't want this logging either.

File: sky/engine/public/sky/sky_headless.h [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+#include <memory>
+
+#include "sky/engine/public/platform/WebString.h"
+#include "sky/engine/wtf/OwnPtr.h"
+#include "sky/engine/wtf/text/WTFString.h"
+
+namespace blink {
+class DartController;
+class DartLibraryProvider;
```
Comment (abarth) [inline sky/engine/public/sky/sky_headless.h]: Do we need this?

File: sky/engine/public/sky/sky_headless.h [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef SKY_ENGINE_PUBLIC_SKY_SKY_HEADLESS_H_
+#define SKY_ENGINE_PUBLIC_SKY_SKY_HEADLESS_H_
+
+#include <memory>
```
Comment (abarth) [inline sky/engine/public/sky/sky_headless.h]: Do we use this?

File: sky/engine/public/sky/sky_headless.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+#include "sky/engine/public/platform/WebString.h"
```
Comment (abarth) [inline sky/engine/public/sky/sky_headless.h]: Ditto

File: sky/packages/updater/BUILD.gn [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  inputs = [
+    main_dart,
+  ]
+  outputs = [
+    snapshot,
+  ]
+
+  sky_snapshot_dir =
+      get_label_info("//sky/tools/sky_snapshot($host_toolchain)", "root_out_dir")
+#  script = rebase_path("$sky_snapshot_dir/sky_snapshot", root_build_dir)
```
Comment (abarth) [inline sky/packages/updater/BUILD.gn]: Remove commented out gn code?

File: sky/packages/updater/BUILD.gn [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  ]
+
+  # TODO(mpcomplete): this is hacky. why can't I just invoke sky_snapshot
+  # directly?
+  asset_base = "//sky/packages/material_design_icons/lib/icons"
+  script = "//sky/tools/sky_build.py"
+  args += [
+    "--asset-base", rebase_path(asset_base, root_build_dir),
+    "--compiler", rebase_path("$sky_snapshot_dir/sky_snapshot", root_build_dir),
+    "--output-file", "/dev/null",
```
Comment (abarth) [inline sky/packages/updater/BUILD.gn]: :(

File: sky/shell/android/update_service_android.cc [hunk 8]
```diff
@@ -21,18 +25,37 @@ bool RegisterUpdateService(JNIEnv* env) {
...
 }
 
+void UpdateTaskAndroid::Start() {
+  Shell::Shared().ui_task_runner()->PostTask(
+      FROM_HERE, base::Bind(&UpdateTaskAndroid::RunDartOnUIThread,
+                            base::Unretained(this)));
+}
+
+void UpdateTaskAndroid::RunDartOnUIThread() {
+  LOG(INFO) << "RunDart";
```
Comment (abarth) [inline sky/shell/android/update_service_android.cc]: Too much logging.

**Comment by abarth** [comment]
This looks pretty good.  You might want to take a pass through cleaning up the extra logging statements and such.  Also the BUILD.gn file that generates the snapshot is way more involved than necessary.  The reason you can't call sky_snapshot directly is that gn requires actions to be written in python.  You should be able to create a python wrapper that calls into sky_snapshot and skips all the extra stuff you add to add to make sky_build work.

**Comment by mpcomplete** [comment]
Cleaned up the logging and extra headers. Added a run_sky_snapshot.py that invokes sky_snapshot with the args I need.

File: sky/engine/core/script/dart_controller.cc [hunk 9]
```diff
@@ -106,6 +106,15 @@ void DartController::RunFromSnapshot(
       base::Bind(&DartController::DidLoadSnapshot, weak_factory_.GetWeakPtr()));
 }
 
+void DartController::RunFromSnapshotBuffer(const uint8_t* buffer, size_t size) {
```
Comment (abarth) [inline sky/engine/core/script/dart_controller.cc]: Is there a reason RunFromSnapshot doesn't eventually call RunFromSnapshotBuffer?

**Comment by abarth** [comment]
LGTM

File: sky/engine/core/script/dart_controller.cc (hunk 9 above)
Comment (mpcomplete) [inline sky/engine/core/script/dart_controller.cc]: I think I could change it do so, but it didn't seem necessary. Also, RunFromSnapshot assumes Dart_CurrentIsolate() is null, which is not the case for my code. Not sure how much that matters.

**Comment by abarth** [comment]
Please squash commits before landing.

**Pull request merged**